eth/stagedsync/stage_tevm.go

```diff
@@ -45,22 +45,81 @@ func StageTranspileCfg(
 	}
 }
 
-func transpileBatch(logPrefix string, s *StageState, fromBlock uint64, toBlock uint64, tx ethdb.RwTx, batch ethdb.DbWithPendingMutations, cfg TranspileCfg, useExternalTx bool, quitCh <-chan struct{}) error {
-	logEvery := time.NewTicker(logInterval)
-	defer logEvery.Stop()
+func SpawnTranspileStage(s *StageState, tx ethdb.RwTx, toBlock uint64, quit <-chan struct{}, cfg TranspileCfg) error {
+	// panics! tx == nil
+
+	prevStageProgress, errStart := stages.GetStageProgress(tx, stages.Execution)
+	if errStart != nil {
+		return errStart
+	}
+
+	var to = prevStageProgress
+	if toBlock > 0 {
+		to = min(prevStageProgress, toBlock)
+	}
+
+	if to <= s.BlockNumber {
+		s.Done()
+		return nil
+	}
 
 	stageProgress := uint64(0)
-	logBlock := stageProgress
 	logTime := time.Now()
 
+	logPrefix := s.state.LogPrefix()
+	log.Info(fmt.Sprintf("[%s] Contract translation", logPrefix), "from", s.BlockNumber, "to", to)
+
+	var lastValue []byte
+	var err error
+	lastKey := dbutils.EncodeBlockNumber(s.BlockNumber + 1)
+
+	excludedAddress := common.Address{}
+	excludedAddress[len(excludedAddress)-1] = 1
+	empty := common.Address{}
+
+	observedAddresses := map[common.Address]struct{}{
+		empty:           {},
+		excludedAddress: {},
+	}
+	observedCodeHashes := map[common.Hash]struct{}{}
+
+	for lastKey != nil {
+		lastKey, lastValue, stageProgress, logTime, err = transpileBatch(logPrefix, s, lastKey, lastValue, to, cfg, tx, observedAddresses, observedCodeHashes, stageProgress, logTime, quit)
```

Comment:
After commit - db must be in consistent state - then indicator of progress must be `stageProgress` instead of `lastKey`

Reply:
It should be, however we are processing many events for each block. So stateProgress >> current block, and we have no way to resume by stateProgress but we can by lasyKey == lastProcessedBlock

Reply:
after commit - need be able to reboot Erigon - and after reboot it must start from correct position

Reply:
after reboot you have only “stage progress” digit.